csi/controller_server.go

```diff
@@ -732,23 +732,49 @@ func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi
 	if existVol.State != string(types.VolumeStateDetached) {
 		return nil, status.Errorf(codes.FailedPrecondition, "Invalid volume state %v for expansion", existVol.State)
 	}
-	volumeSize, err := strconv.ParseInt(existVol.Size, 10, 64)
+	existingSize, err := strconv.ParseInt(existVol.Size, 10, 64)
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, err.Error())
 	}
+
+	// we can only call the longhorn expansion api if the requested size is larger than the volume size
+	// since longhorn treats the size differently than kubernetes, in kubernetes capacity is a request
+	// to ensure that the volume has at least that amount of capacity.
 	requestedSize := req.CapacityRange.GetRequiredBytes()
-	if requestedSize <= volumeSize {
-		return nil, status.Errorf(codes.OutOfRange, "The requested size %v is smaller than or the same as the size %v of volume %v", requestedSize, volumeSize, existVol.Name)
+	if requestedSize > existingSize {
+		if existVol, err = cs.apiClient.Volume.ActionExpand(existVol, &longhornclient.ExpandInput{
+			Size: strconv.FormatInt(requestedSize, 10),
+		}); err != nil {
+			return nil, status.Errorf(codes.Internal, err.Error())
+		}
+	}
+
+	// kubernetes doesn't support volume shrinking and the csi spec specifies to return true
+	// in the case where the current capacity is bigger or equal to the requested capacity
+	// that's why we return the volumeSize below instead of the requested capacity
+	volumeExpansionComplete := func(vol *longhornclient.Volume) bool {
+		engineReady := false
+		if len(vol.Controllers) > 0 {
+			engine := vol.Controllers[0]
+			engineSize, _ := strconv.ParseInt(engine.Size, 10, 64)
+			engineReady = engineSize >= requestedSize && !engine.IsExpanding
+		}
+		size, _ := strconv.ParseInt(vol.Size, 10, 64)
+		return size >= requestedSize && engineReady
+	}
+
+	if !cs.waitForVolumeState(req.VolumeId, "volume expansion", volumeExpansionComplete, false, false) {
+		return nil, status.Errorf(codes.DeadlineExceeded, "Expanding volume %s existing capacity %v requested capacity %v failed",
+			req.GetVolumeId(), existingSize, requestedSize)
 	}
```

Comment:
Is it a problem if it takes longer than 2 minutes to finish the expansion?
Does it timeout on the Kubernetes side? Does Kubernetes retry the expansion call?

Reply:
Not a problem, all csi calls needs to be idempotent.
It's important to wait for the expansion to succeed otherwise you will have signaled that the system has updated the size, which will lead to the external-resizer updating the PV size. 

Now imagine what would happen if there is a crash in the middle of any of this. Now the PV and actual volume size might be out of sync. So by waiting we ensure that longhorn and the PV size are in sync.

If longhorn has already done the update, but the PV update failed then this call will be retried since the PVC.requestedSize > PV.size. That's why it's important that this call is idempotent.

Reply:
Step all the way towards: 
https://github.com/kubernetes-csi/external-resizer/blob/372f7514451bd6f54023e8df5860f606cc73f938/pkg/controller/controller.go#L270
https://github.com/kubernetes-csi/external-resizer/blob/372f7514451bd6f54023e8df5860f606cc73f938/pkg/controller/controller.go#L437
https://github.com/kubernetes-csi/external-resizer/blob/372f7514451bd6f54023e8df5860f606cc73f938/pkg/resizer/csi_resizer.go#L177

final csi call:
https://github.com/kubernetes-csi/external-resizer/blob/372f7514451bd6f54023e8df5860f606cc73f938/pkg/csi/client.go#L122

Reply:
Thank you for the explanation